runtime/recipe/type-checker.js

```diff
@@ -46,7 +46,9 @@ class TypeChecker {
     // TODO: direction?
     let type;
     if (leftType.isVariable) {
-      leftType.variable.resolution = rightType;
+      if (!rightType.isVariable) {
+        leftType.variable.resolution = rightType;
+      }
```

Comment:
This is incorrect. When comparing variables we must always establish equality between them or more complicated type constructions won't resolve.

Reply:
reverted this change and added a test we discussed offline.
everything else seems to be working fine.
PTAL